### PR TITLE
Molden format writer for ccwrite

### PR DESCRIPTION
**Pull request by gaursagar**
Adding initial support for writing Molden format files.

### CODE REVIEW

File: src/cclib/io/ccio.py [hunk 1]
```diff
@@ -363,6 +364,8 @@ def _determine_output_format(outputtype, outputdest):
             outputclass = cmlwriter.CML
         elif outputtype.lower() == 'xyz':
             outputclass = xyzwriter.XYZ
+        elif outputtype.lower() == 'molden':
```
Comment (langner) [inline src/cclib/io/ccio.py]: Would it make sense to use a dictionary for this mapping instead of all the conditions?

Comment (langner) [inline src/cclib/io/ccio.py]: Could you also add a unit test for this function?

File: src/cclib/io/moldenwriter.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: How about a unit test?

File: src/cclib/io/ccio.py (hunk 1 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: Dictionary sounds like a better idea. In future there might be a lot more writers.

Comment (gaursagar) [inline src/cclib/io/ccio.py]: https://github.com/cclib/cclib/pull/385/commits/96d30ccd04b43b73e671a7c468711c8f3b22b6f2 Moved the mappings to a dictionary.

Comment (gaursagar) [inline src/cclib/io/ccio.py]: https://github.com/cclib/cclib/pull/385/commits/7095913790a7027794a1f8ac658c26d980ab154f I added a unit test for the function.

File: src/cclib/io/ccio.py [hunk 3]
```diff
@@ -355,29 +356,33 @@ def _determine_output_format(outputtype, outputdest):
...
-            outputclass = cjsonwriter.CJSON
-        elif outputtype.lower() == 'cml':
-            outputclass = cmlwriter.CML
-        elif outputtype.lower() == 'xyz':
-            outputclass = xyzwriter.XYZ
+        try:
+            outputclass = outputclasses[outputtype.lower()]
+        except:
+            # Determine the outputclass using outputdest in else part.
+            pass
```
Comment (TDI-tenderholt) [inline src/cclib/io/ccio.py]: Shouldn't this raise an exception like below?

File: src/cclib/io/ccio.py [hunk 4]
```diff
@@ -355,29 +356,33 @@ def _determine_output_format(outputtype, outputdest):
...
-        elif extension.lower() == '.cml':
-            outputclass = cmlwriter.CML
-        elif extension.lower() == '.xyz':
-            outputclass = xyzwriter.XYZ
-        else:
+        try:
+            print (extension[1:])
+            outputclass = outputclasses[extension[1:]]
+        except:
             raise ValueError
```
Comment (TDI-tenderholt) [inline src/cclib/io/ccio.py]: The exception should have more info (e.g. "Unknown/unsupported extension.").

File: src/cclib/io/moldenwriter.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        the file path containing forward slash or backward slash - which is valid
+        in Windows.
+        """
+        head, tail = ntpath.split(path)
+        return tail or ntpath.basename(head)
+
+    def _coords_from_ccdata(self, index):
+        """Create [Atoms] section using geometry at the given index."""
+
+        element_list = [self.pt.element[Z] for Z in self.ccdata.atomnos]
```
Comment (TDI-tenderholt) [inline src/cclib/io/moldenwriter.py]: Isn't this already calculated in ```filewriter.Writer.__init__()```, right after where ```self.pt``` is set?

File: src/cclib/io/moldenwriter.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        atomcoords = self.ccdata.atomcoords[index]
+        atomnos = self.ccdata.atomnos
+        nos = range(1, self.ccdata.natom)
+        
+        # element_name number atomic_number x y z
+        atom_template = '{:3s} {:3d} {:3d} {:15.10f} {:15.10f} {:15.10f}'
```
Comment (TDI-tenderholt) [inline src/cclib/io/moldenwriter.py]: This is already clear, but was thinking you could also be more explicit by using field names:

```atom_template = '{element:3s} {no:3d} {atomno:3d} {x:15.10f} {y:15.10f} {z:15.10f}'```

File: src/cclib/io/moldenwriter.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        # shell_label number_of_primitives 1.00
+        # exponent_primitive_1 contraction_coefficient_1 (contraction_coefficient_1)
+        # ...
+        # empty line
+        # atom_sequence__number2 0
+        gbasis = self.ccdata.gbasis
+        block = []
+        label_template = '{:s} {:5d} 1.0'
+        basis_template = '  {:15.10f} {:15.10f}'
+        block = []
```
Comment (TDI-tenderholt) [inline src/cclib/io/moldenwriter.py]: Similarly, you could be more explicit here with field names.

File: src/cclib/io/ccio.py [hunk 8]
```diff
@@ -356,33 +356,33 @@ def _determine_output_format(outputtype, outputdest):
     #  1. outputtype
     #  2. outputdest
 
+    outputclasses = {
```
Comment (langner) [inline src/cclib/io/ccio.py]: Perhaps make this a module constant?

File: test/io/testccio.py [hunk 9]
```diff
@@ -111,6 +111,25 @@ def test_multi_url_io(self):
...
 
 
+class _determine_output_formatTest(unittest.TestCase):
+
+    def setUp(self):
+        self._determine_output_format = cclib.io.ccio._determine_output_format
+
+    def test_outputclass(self):
+        """Does the function determine output class as expected"""
+        absdir = os.path.dirname(os.path.abspath(__file__))
```
Comment (langner) [inline test/io/testccio.py]: Since the function only looks at the extension, is there any need to play around with directories here?

File: src/cclib/io/ccio.py [hunk 10]
```diff
@@ -355,29 +356,33 @@ def _determine_output_format(outputtype, outputdest):
     #  1. outputtype
```
Comment (langner) [inline src/cclib/io/ccio.py]: What is the purpose of these comments?

File: src/cclib/io/ccio.py [hunk 11]
```diff
@@ -286,7 +287,7 @@ def ccwrite(ccobj, outputtype=None, outputdest=None, terse=False , returnstr=Fal
 
     Inputs:
         ccobj - Either a job (from ccopen) or a data (from job.parse()) object
-        outputtype - The output format (should be one of 'cjson', 'cml', 'xyz')
+        outputtype - The output format (should be one of 'cjson', 'cml', 'xyz', 'molden')
```
Comment (langner) [inline src/cclib/io/ccio.py]: I don't think you need to enumerate the keys here, just maybe specify that this is expected to be string. Is it?

File: src/cclib/io/ccio.py [hunk 12]
```diff
@@ -355,29 +356,33 @@ def _determine_output_format(outputtype, outputdest):
...
             raise ValueError
-        if extension.lower() in ('.cjson', '.json'):
-            outputclass = cjsonwriter.CJSON
-        elif extension.lower() == '.cml':
-            outputclass = cmlwriter.CML
-        elif extension.lower() == '.xyz':
-            outputclass = xyzwriter.XYZ
-        else:
+        try:
+            print (extension[1:])
```
Comment (langner) [inline src/cclib/io/ccio.py]: If the leftover from debugging? :)

File: src/cclib/io/ccio.py [hunk 13]
```diff
@@ -345,7 +346,7 @@ def _determine_output_format(outputtype, outputdest):
 
     Inputs:
       outputtype - a string corresponding to the file type
-        (one of cjson/json, cml, xyz)
+        (one of cjson/json, cml, xyz, molden)
```
Comment (langner) [inline src/cclib/io/ccio.py]: I don't this you need to enumerate the keys in the docstring.

File: src/cclib/io/moldenwriter.py [hunk 14]
```diff
@@ -76,6 +76,21 @@ def _gto_from_ccdata(self):
 
         return block
 
+    def _scfconv_from_ccdata(self):
+        """Create [SCFCONV] section using gbasis."""
+
+        # scf-first    1 THROUGH   12
```
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: No. Molden seems to handles spaces quite gracefully.

File: src/cclib/io/ccio.py (hunk 10 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: This is an artefact from while I was coding; should not have been a part of the code. I'll remove all these in a commit.

File: src/cclib/io/ccio.py (hunk 12 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: Yes. I forgot to clean up this particular file before committing. Will do it in the next commit.

File: src/cclib/io/moldenwriter.py (hunk 14 above)
Comment (langner) [inline src/cclib/io/moldenwriter.py]: So is there any reasons have some many spaces here?

Also, please cover the edge cases in the unit test.

File: src/cclib/io/ccio.py (hunk 10 above)
Comment (langner) [inline src/cclib/io/ccio.py]: Please squash that commit into the original.

**Comment by gaursagar** [comment]
After the latest commit, we have a working Molden writer. The written files can be opened using the Molden program ( tested on [`data/GAMESS/basicGAMESS-US2014`](https://github.com/cclib/cclib/tree/master/data/GAMESS/basicGAMESS-US2014)). 
Now, after making a diff based test suite and resolving the code review suggestions this PR can be closed.

**Comment by langner** [comment]
> Now, after making a diff based test suite and resolving the code review suggestions this PR can be closed.

You mean merged :)

File: src/cclib/io/moldenwriter.py [hunk 15]
```diff
@@ -37,15 +37,15 @@ def pathname(self, path):
...
+        elements = self.elements
         atomcoords = self.ccdata.atomcoords[index]
         atomnos = self.ccdata.atomnos
         nos = range(self.ccdata.natom)
 
         # element_name number atomic_number x y z
         atom_template = '{:2s} {:5d} {:2d} {:12.6f} {:12.6f} {:12.6f}'
         block = []
-        for element, no, atomno, (x, y, z) in zip(element_list, nos, atomnos,
+        for element, no, atomno, (x, y, z) in zip(elements, nos, atomnos,
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: Why not just self.elements here?

File: src/cclib/io/ccio.py [hunk 16]
```diff
@@ -345,7 +354,6 @@ def _determine_output_format(outputtype, outputdest):
 
     Inputs:
       outputtype - a string corresponding to the file type
-        (one of cjson/json, cml, xyz)
       outputdest - a filename string or file handle
     Returns:
```
Comment (langner) [inline src/cclib/io/ccio.py]: Could you also add a "Raises" part explaining the exception raised?

File: src/cclib/io/ccio.py (hunk 3 above)
Comment (langner) [inline src/cclib/io/ccio.py]: Please add this case to the unit test.

File: src/cclib/io/moldenwriter.py (hunk 5 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Addressed. Now self.elements is being used.

File: src/cclib/io/ccio.py (hunk 16 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: Added.

File: src/cclib/io/ccio.py (hunk 4 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: Now a `KeyError` is being raised with a custom messege `Unsupported Extension <ext>`.

File: src/cclib/io/ccio.py (hunk 3 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: KeyError with a message is being raised, and the case is being covered in the unit test.

File: src/cclib/io/moldenwriter.py (hunk 6 above)
Comment (langner) [inline src/cclib/io/moldenwriter.py]: How about this?

File: src/cclib/io/moldenwriter.py (hunk 7 above)
Comment (langner) [inline src/cclib/io/moldenwriter.py]: And this?

File: src/cclib/io/moldenwriter.py (hunk 14 above)
Comment (langner) [inline src/cclib/io/moldenwriter.py]: So, is there any reason to keep all the spaces?

File: src/cclib/io/ccio.py (hunk 10 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: Done.

File: src/cclib/io/moldenwriter.py (hunk 14 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Spacing does not matter. These spacing are however in accordance to how Molden program prints spaces.

File: test/io/testccio.py (hunk 9 above)
Comment (gaursagar) [inline test/io/testccio.py]: Changed. Now the test passes the extension as a string.
And there is a test to send an unsupported extension.

File: src/cclib/io/ccio.py (hunk 8 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: Done.

File: src/cclib/io/ccio.py (hunk 13 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: Updated.

File: src/cclib/io/ccio.py [hunk 17]
```diff
@@ -345,39 +354,37 @@ def _determine_output_format(outputtype, outputdest):
...
-            outputclass = cjsonwriter.CJSON
-        elif outputtype.lower() == 'cml':
-            outputclass = cmlwriter.CML
-        elif outputtype.lower() == 'xyz':
-            outputclass = xyzwriter.XYZ
+        try:
+            outputclass = outputclasses[outputtype.lower()]
+        except KeyError as e:
+            e.args = ('Unsupported Extension %s' % e.args[0],)
+            raise
```
Comment (langner) [inline src/cclib/io/ccio.py]: Better to re-raise the exception explicitly. Also, after looking at this code it seems like a custom exception would make sense here. UnknownOutputFormatError?

File: src/cclib/io/moldenwriter.py [hunk 18]
```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2016, the cclib development team
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: 2017

File: src/cclib/io/moldenwriter.py [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+#
+# This file is part of cclib (http://cclib.github.io) and is distributed under
+# the terms of the BSD 3-Clause License.
+
+"""A writer for MOLDEN format files."""
+
+import os.path
+import ntpath
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: Just stick with os.path

File: src/cclib/io/moldenwriter.py [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+
+from . import filewriter
+from cclib.parser import utils
+
+
+class MOLDEN(filewriter.Writer):
+    """A writer for chemical JSON (MOLDEN) files."""
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: remove chemical JSON

File: src/cclib/io/moldenwriter.py [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    def __init__(self, ccdata, *args, **kwargs):
+        """Initialize the chemical JSON writer object.
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: Remove chemical JSON

File: src/cclib/io/moldenwriter.py [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+
+        Inputs:
+          ccdata - An instance of ccData, parsed from a logfile.
+        """
+        # Call the __init__ method of the superclass
+        super(MOLDEN, self).__init__(ccdata, *args, **kwargs)
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: You don't need to do this since you're not adding anything to the initializer.

File: src/cclib/io/moldenwriter.py (hunk 6 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: I would argue that, having field names would look less cleaner while formatting the string templates. Additionally, there are comments specifying what fields are to be filled in the string.

```Python
# element_name number atomic_number x y z
atom_template = '{element:2s} {no:5d} {atomno:2d} {x:12.6f} {y:12.6f} {z:12.6f}'
...
block.append(atom_template.format(element=element, no=no, atomno=atomno,
                                  x=x, y=y, z=z))
```

File: test/io/testccio.py [hunk 23]
```diff
@@ -111,6 +111,22 @@ def test_multi_url_io(self):
...
+    def setUp(self):
+        self._determine_output_format = cclib.io.ccio._determine_output_format
+
+    def test_outputclass(self):
+        """Does the function determine output class as expected"""
+        outputtype = "xyz"
+        outputdest = "file.xyz"
+        self.assertEqual(self._determine_output_format(outputtype, outputdest),
+                         cclib.io.xyzwriter.XYZ)
+        # Must raise a KeyError for unsuported extensions
```
Comment (langner) [inline test/io/testccio.py]: nit: end sentences with periods

File: src/cclib/io/moldenwriter.py [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+                    block.append('{:4d}  {:10.6f}'.format(k + 1, mocoeff))
+
+            spin = 'Beta'
+
+        return block
+
+    def generate_repr(self):
+        """Generate the MOLDEN representation of the logfile data."""
+
+        molden_block = ['[Molden Format]']
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: Having 'block' in the name is a little misleading, since this is a list of lines.

File: src/cclib/io/moldenwriter.py [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+
+        # Coordinates for the Electron Density/Molecular orbitals
+        # [Atoms] (Angs|AU)
+        unit = "Angs"
+        molden_block.append('[Atoms] %s' % unit)
+        # Last set of coordinates for geometry optimization runs.
+        index = -1
+        molden_block.extend(self._coords_from_ccdata(index))
+
+        if hasattr(self.ccdata, 'gbasis'):
```
Comment (langner) [inline src/cclib/io/moldenwriter.py]: I'm curious - how does Molden react if one of these sections is missing? You probably want to address this in the tests you will write as well.

Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: `[GTO]` and `[MO]` sections should be both present together. Coordinates are mandatory.

File: src/cclib/io/moldenwriter.py (hunk 24 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Renamed `block` to `lines`

File: test/io/testccio.py (hunk 23 above)
Comment (gaursagar) [inline test/io/testccio.py]: Done.

File: src/cclib/io/moldenwriter.py (hunk 22 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Removed the `__init__` function.

File: src/cclib/io/moldenwriter.py (hunk 21 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Done.

File: src/cclib/io/moldenwriter.py (hunk 20 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Done.

File: src/cclib/io/moldenwriter.py (hunk 19 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Refactored the `title` extraction module to use only `os.path.basename` and `os.path.splitext`.

File: src/cclib/io/moldenwriter.py (hunk 18 above)
Comment (gaursagar) [inline src/cclib/io/moldenwriter.py]: Done.

File: src/cclib/io/ccio.py (hunk 17 above)
Comment (gaursagar) [inline src/cclib/io/ccio.py]: This part has been refactored to check if the extension is supported; and if not, raise a custom exception `UnknownOutputFormatError`.

File: src/cclib/io/ccio.py (hunk 16 above)
Comment (TDI-tenderholt) [inline src/cclib/io/ccio.py]: Looks like it now raises ```UnknownOutputFormatError```, not ```KeyError```.

Comment (gaursagar) [inline src/cclib/io/ccio.py]: It was updated to `UnknownOutputFormatError ` based on the later reviews.

Comment (TDI-tenderholt) [inline src/cclib/io/ccio.py]: Docstring needs to be updated (see line 366).
